sympy/polys/matrices/sdm.py

```diff
@@ -95,6 +278,31 @@ def __rmul__(a, b):
             return NotImplemented
 
     def matmul(A, B):
+        """
+        Performs matrix multiplication of two SDM matrices
+
+        Parameters
+        ==========
+
+        A, B: SDM to multiply
+
+        Returns
+        =======
+
+        SDM
+            SDM after multiplication
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> A = SDM({0:{1: ZZ(2)}, 1:{0:ZZ(1)}}, (2, 2), ZZ)
+        >>> B = SDM({0:{0:ZZ(2), 1:ZZ(3)}, 1:{0:ZZ(4)}}, (2, 2), ZZ)
+        >>> A.matmul(B)
+        {0: {0: 8}, 1: {0: 2, 1: 3}}
```

Comment:
For each of these methods it would be good to point out that they are also used by the operators so e.g. this could show that you get the same from `A*B`.

Reply:
These docstrings should also explain when the domains are are required to match or what constraints there are on the shapes of the matrices for the operation to be defined.

Reply:
You actually don't get the same from `A * B`. 

>>> A * B
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'SDM' and 'SDM'

This is what you get with A*B

Reply:
I think ` __add__(A, B)` needs to be implemented for SDM class

Reply:
Oh yeah, I think it's `A @ B` or `__matmul__`. Is that not implemented for `SDM`?

Reply:
Perhaps it is better to use `*` anyway. Either way `SDM` and `DDM` should be made consistent so can you compare how `DDM` works?

Reply:
Consistent in what ways? You want both to have the same functionality right, is that what you're saying

Reply:
By the way `*` does not work in DDM either
```python
A = DDM([[ZZ(1), ZZ(2)], [ZZ(2), ZZ(3)]], (2, 2), ZZ)
B = DDM([[ZZ(3), ZZ(4)], [ZZ(1), ZZ(2)]], (2, 2), ZZ)
A * B
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'DDM' and 'DDM
```

Reply:
I mean that when using methods like `A * B` or `A.matmul(B)` etc then all behaviour should be exactly the same whether `A` and `B` are both `DDM` or both `SDM`. These two classes are supposed to be completely interchangeable in that sense. So if you are changing any part of the behaviour of `SDM` you should check that `DDM` behaves in exactly the same way.

Reply:
> I'll have to make `__mul__ ` in DDM consistent with definition in DomainMatrix

Note that `DDM` and `DomainMatrix` do not need to be consistent as they are not intended to be interchangeable. It is `DDM` and `SDM` that need to be consistent. For now I think that `SDM` should just use `@` for matrix multiplication just like `DDM` does. We can consider changing that in future but these two classes need to remain equivalent.